# operaciones con arrays
Existen diferentes operaciones que se pueden usar para los arras de numpy

### duplicar los valores de una lista
Dada la lista $\rightarrow$ `lista = [1,2]`

*multiplicar los valores de la lista por 2*
* sin numpy: `lista = [1, 2]` $\rightarrow$ `lista * 2` $\rightarrow$ `[1,2,1,2]`
  * esto **NO es lo que buscamos**. Aquí lo que se hace es concatenar un ser extra del mismo array
    * python entiende que se quieren multiplicar los valores. Pero de manera literal, no matemática   
/
  * para hacerlo sin nunmpy lo que tendríamos que hacer es una list comprehension que multiplique cada número 
  
    `[i*2 for i in lista]`
* con numpy: esta tarea es más "sencilla". Dada la lista definida en un array, simplemente lo multiplicamos directamente * 2, tal como habíamos intentado anteriormente




In [2]:
import numpy as np

In [17]:
lista= [1,2]
array_lista = np.array(lista) # transformar la lista en un array

#multiplicación
array_lista * 2

print(f'  sin numpy, concatena los elementos con el mismo set: lista * 2 = {lista *2}\n',
      f' con numpy opera los elementos: array_lista * 2 = {array_lista *2}')


  sin numpy, concatena los elementos con el mismo set: lista * 2 = [1, 2, 1, 2]
  con numpy opera los elementos: array_lista * 2 = [2 4]


In [35]:
# multiplicar elementos dentro de un vector
vector = np.array([np.linspace(1,4,4)], dtype= int)

print(f' vector original: {vector} vector *2 : {vector*2}')

 vector original: [[1 2 3 4]] vector *2 : [[2 4 6 8]]


In [38]:
# multiplicar elementos dentro de una matriz
matriz = np.array([np.linspace(1,4,4), np.linspace(5,8,4)], dtype= int)

print(f' matriz original: \n {matriz} \n *** \n matriz *2:  \n {matriz*2}')

 matriz original: 
 [[1 2 3 4]
 [5 6 7 8]] 
 *** 
 matriz *2:  
 [[ 2  4  6  8]
 [10 12 14 16]]


In [61]:
# multiplicar elementos dentro de un tensor
tensor = np.array([
                    [np.linspace(1,4,4), np.linspace(5,8,4)],
                    [np.linspace(9,12,4), np.linspace(13,16,4)],
                    [np.linspace(17,20,4), np.linspace(21,24,4)],
                  ], dtype= int)

#print(f' tensor original: \n {tensor}\n *** \n tensor *2 : \n{tensor*2}')

In [60]:
# elevar un vector al cuadrado
cuadrado = vector **2
cuadrado

array([[ 1,  4,  9, 16]])

### operaciones matemáticas entre vectores , entre matrices , entre tensores

cada uno de las operaciones hechas entre cada uno de los array de distintas dimensiones puede aplicarse a los demás
 

In [83]:
#definimos los array para operarlos entre si

#vectores
vector_1 = np.arange(1,5)
vector_2 = np.arange(5,9)

#matrices
matriz_1 = vector_1.reshape(2,2)
matriz_2 = vector_2.reshape(2,2)

#tensores
tensor_1 =  np.array([matriz, matriz])
tensor_2 = np.array([matriz, matriz])


In [86]:

suma_vector = vector_1 + vector_2
resta_matriz = matriz_1 - matriz_2
mult_tensor = tensor_1 * tensor_2
div_matriz = matriz_1 / matriz_2

print(f' suma entre vectores: {suma_vector}\n',
      f'resta entre matrices:\n {resta_matriz}\n ***\n',
      f' multiplicación entre tensores: \n{mult_tensor}\n***\n',
      f'división entre matrices:\n {div_matriz}')


 suma entre vectores: [ 6  8 10 12]
 resta entre matrices:
 [[-4 -4]
 [-4 -4]]
 ***
  multiplicación entre tensores: 
[[[ 1  4]
  [ 9 16]]

 [[ 1  4]
  [ 9 16]]]
***
 división entre matrices:
 [[0.2        0.33333333]
 [0.42857143 0.5       ]]


## producto interno / producto punto entre matrices

 una operación importante es la de punto por punto, la podemos hacer con la función `np.matmul()`

 **cuando ambas matrices no son cuadradas**, no es posible hacer el producto interno entre ellas de manera directa.
  * para poder hacer un producto punto entre matrices no cuadradas *(producto interno)* es necesario transponer una de las matrices para que las filas de una matriz tengan la misma cantidad de valores que las columnas de la otra 
    * La razón de esto es porque el número de filas deben hacer match con el número de columnas de la otra matriz

El producto interno también puede hacerse entre una matriz y un vector, solo debe tener la misma cantidad de columnas que la cantidad de filas de la matriz

Otra manera de calcular el producto interno entre matrices es con el operador `@` siendo: `matriz_1 @ matriz_2`

**NOTA** Al calcular el producto interno entre matrices cuadradas, no es necesario transponer, si se hace es posible que el resultado esté mal

In [93]:
np.matmul(matriz_2, np.array([1,2]).T)

array([17, 23])

In [88]:
#con matrices cuadradas no es necesario transponer
np.matmul(matriz_1, matriz_2)

array([[19, 22],
       [43, 50]])

In [94]:
producto_punto = matriz_1 @ matriz_2
producto_punto

array([[19, 22],
       [43, 50]])

### producto punto != multiplicación entre matrices

La multiplicación entre matrices consiste en una multiplicación entre valores ubicados en la misma posición (como en la suma) 

Otra manera de calcular el producto punto es usando `matriz_1.dot(matriz_2.T)` 

### producto punto != multiplicación entre matrices 

In [51]:
matriz_a = np.arange(0,6).reshape(2,3)
matriz_b = np.arange(6,12).reshape(2,3)

print(matriz_a, "\n\n", matriz_b )

[[0 1 2]
 [3 4 5]] 

 [[ 6  7  8]
 [ 9 10 11]]


In [49]:
dot = matriz_a.dot(matriz_b.T)
matmul = np.matmul(matriz_a, matriz_b.T)

mult = matriz_a * matriz_b

print(f' producto punto: \n usando .dot: \n {dot} \n usando matmul: \n{matmul}',
      f' \n multiplicación entre matrices: \n {mult}')

 producto punto: 
 usando .dot: 
 [[ 23  32]
 [ 86 122]] 
 usando matmul: 
[[ 23  32]
 [ 86 122]]  
 multiplicación entre matrices: 
 [[ 0  7 16]
 [27 40 55]]


#### multiplicación entre matrices
fila1 * fila 1

elemento (1,1) $\rightarrow 0 * 6 = 0$

elemento (1,2) $\rightarrow 1 * 7 = 7$

elemento (1,3) $\rightarrow 2 * 8 = 16$

*lo mismo con fila2 x fila 2*

### producto interno entre matrices

elemento (1,1) $\rightarrow 0*6 + 1*7 + 2*8 = 23$ 
*  fila1 * columna 1

elemento (1,2) $\rightarrow 0*9 + 1*10 + 2*11 = 32$ 
* fila 1 * columna 2

elemento (2,1) $\rightarrow 3*6 + 4*7 + 5*8 = 86$ 
* fila 2 * columna 1

elemento (2,2) $\rightarrow 3*9 + 4*10 + 5*11 = 122$ 
* fila 2 * columna 2